File: plugins/python/template/base/app/middleware/rate_limit_middleware.py
import time

from starlette.middleware.base import BaseHTTPMiddleware

from app.core.config import settings
from app.core.exception import TooManyRequestsException
from app.core.redis_client import redis_client


class RateLimitMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        if redis_client.redis:
            return await call_next(request)

        client_id = request.client.host
        api_key = request.headers.get(settings.API_KEY_HEADER)
        if api_key:
            client_id = f"api_key:{api_key}"

        rate_key = f"rate_limit:{client_id}:{int(time.time() / 60)}"
        current = await redis_client.get(rate_key)
        if current and int(current) >= settings.RATE_LIMIT_PER_MINUTE:
            raise TooManyRequestsException()

        if current and redis_client.redis:
            await redis_client.redis.incr(rate_key)
        else:
            await redis_client.set(rate_key, "1", expire=60)
        response = await call_next(request)
        return response

File: plugins/node/template/base/src/middleware/rate-limit.middleware.ts
import { FastifyReply, FastifyRequest } from "fastify";
import redis from "../config/redis";
import { env } from "../config";
import { ERROR_RATE_LIMIT_EXCEEDED } from "../constants";


export async function rateLimitMiddleware(
  request: FastifyRequest,
  reply: FastifyReply
) {
  if (!redis) {
    return
  }

  let clientId = request.ip;

  const apiKey = request.headers["x-api-key"];
  if (apiKey) {
    clientId = `api_key:${apiKey}`;
  }

  const now = Math.floor(Date.now() / 60000);
  const rateKey = `rate_limit:${clientId}:${now}`

  try {
    const current = await redis.get(rateKey)

    if (current && parseInt(current) >= (env.RATE_LIMIT_MAX || 100)) {
      return reply.status(429).send({
        success: false,
        error: ERROR_RATE_LIMIT_EXCEEDED
      })
    }

    if (current) {
      await redis.incr(rateKey)
    } else {
      await redis.set(rateKey, "1", "EX", 60);
    }
  } catch (error) {
    request.log.error({ error }, "Rate limit check failed")
  }
}
